### PR TITLE
niLang: Add ni::Nonnull, astl::non_null and make ni::Ptr/WeakPtr/QPtr safer ;

### PR DESCRIPTION
**Pull request by prenaux**
## Formal Notes

This PR fixes:
* RN: niLang: Add `ni::Nonnull` & `astl::non_null`.
* RN: niLang: Safer API for `ni::Ptr`, `ni::QPtr` & `ni::WeakPtr`. Can be made strict with `#define niNoUnsafePtr`.

## Description, Motivation and Context

The goal is to introduce a clear set of safe C++ types that we
can use in our new code. Eventually the older part of the
codebase should be modernised to use these safer alternatives.

- New generic pointer types: `astl::non_null`.

- iUnknown pointer types with a common safe API: `ni::Ptr`, `ni::QPtr` & `ni::WeakPtr`.
  - `is_null` / `has_value`: Check whether the pointer points to an object.
  - `non_null` / `c_non_null`: Return the Nonnull pointer to the object. (Panic if trying to retrieve a Null object)
  - `raw_ptr`: Return the underlying pointer directly without any check.
- The safe API can be enforced with with `#define niNoUnsafePtr` - this will remove the unsafe accessor from ni::Ptr/ni::QPtr.
- Removed ni::MemberPointer and replaced it by ni::Nonnull.

Pointer types relationship, the types in blue are the new ones introduced in this PR:

Screenshot 2022-09-03 at 10 56 34 AM

## How Has This Been Tested?

Command line to run all the tests that I run locally:
```
ham Run_ci
```

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that cause existing functionality to change)
* [ ] Documentation (code docs, comments, or changes to the doc systems)
* [ ] Testing/Build (test coverage or the test/build subsystems themselves)
* [ ] Packaging (adds examples or modifies a release package)

### CODE REVIEW

**Comment by prenaux** [comment]
Current dependencies on/for this PR:
* main
  * **PR #17** <a href="https://app.graphite.dev/github/pr/prenaux/niLang/17" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/prenaux/niLang/17?utm_source=stack-comment).

**Review by Edensan — CHANGES_REQUESTED** [review]
See inline comments

File: sources/niLang/src/API/niLang/Math/MathProb.h
Comment (prenaux) [reply sources/niLang/src/API/niLang/Math/MathProb.h:99]: Because its not supposed to ever hold a reference to it. astl::non_null<T*> is equivalent to T*, Nonnull<T> is equivalent to Ptr<T>.

File: sources/niLang/src/API/niLang/Utils/CollectionImpl.h
Comment (prenaux) [reply sources/niLang/src/API/niLang/Utils/CollectionImpl.h:764]: Yes indeed, this should be astl::non_null<>. I've done only the minimum to make it compile with niNoUnsafePtr.

I created https://github.com/prenaux/niLang/issues/23 to keep track of this.

**Comment by prenaux** [comment]
**Graphite Merge Job**

Current status: 🔄 Waiting for checks to pass

This pull request is currently waiting for CI and mergeability checks to pass before being merged.

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/prenaux/niLang/17?utm_source=stack-comment).


Job Reference: QrsxxOZ6CSlm51a9vKf7

**Pull request merged**
